lib/clientv2/client.go

```diff
@@ -0,0 +1,164 @@
+func (c client) NetworkPolicies(namespace string) NetworkPolicyInterface {
+	return networkPolicies{client: c, namespace: namespace}
+}
+
+// Policies returns an interface for managing policy resources.
+func (c client) GlobalNetworkPolicies() GlobalNetworkPolicyInterface {
+	return globalnetworkpolicies{client: c}
+}
+
+// IPPools returns an interface for managing IP pool resources.
+func (c client) IPPools() IPPoolInterface {
+	return ipPools{client: c}
+}
+
+// Profiles returns an interface for managing profile resources.
+func (c client) Profiles() ProfileInterface {
+	return profiles{client: c}
+}
+
+// HostEndpoints returns an interface for managing host endpoint resources.
+func (c client) HostEndpoints() HostEndpointInterface {
+	return hostEndpoints{client: c}
+}
+
+// WorkloadEndpoints returns an interface for managing workload endpoint resources.
+func (c client) WorkloadEndpoints(namespace string) WorkloadEndpointInterface {
+	return workloadEndpoints{client: c, namespace: namespace}
+}
+
+// BGPPeers returns an interface for managing BGP peer resources.
+func (c client) BGPPeers() BGPPeerInterface {
+	return bgpPeers{client: c}
+}
+
+// IPAM returns an interface for managing IP address assignment and releasing.
+func (c client) IPAM() ipam.Interface {
+	return ipam.NewIPAM(c.Backend, poolAccessor{})
+}
+
+type poolAccessor struct {
+	client *client
+}
+
+func (p poolAccessor) GetEnabledPools(ipVersion int) ([]net.IPNet, error) {
+	pools, err := p.client.IPPools().List(options.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	enabled := []net.IPNet{}
+	for _, pool := range pools.Items {
+		if pool.Spec.Disabled {
+			continue
+		} else if _, cidr, err := net.ParseCIDR(pool.Spec.CIDR); err != nil {
+			enabled = append(enabled, *cidr)
+		}
+	}
+	return enabled, nil
+}
+
+// EnsureInitialized is used to ensure the backend datastore is correctly
+// initialized for use by Calico.  This method may be called multiple times, and
+// will have no effect if the datastore is already correctly initialized.
+//
+// Most Calico deployment scenarios will automatically implicitly invoke this
+// method and so a general consumer of this API can assume that the datastore
+// is already initialized.
+func (c client) EnsureInitialized() error {
+	// Perform datastore specific initialization first.
+	if err := c.Backend.EnsureInitialized(); err != nil {
+		return err
+	}
+
+	// Ensure a cluster GUID is set for the deployment.  We do this
+	// irrespective of how Calico is deployed.
+	kv := &model.KVPair{
+		Key:   model.GlobalConfigKey{Name: "ClusterGUID"},
+		Value: fmt.Sprintf("%v", hex.EncodeToString(uuid.NewV4().Bytes())),
+	}
+	if _, err := c.Backend.Create(kv); err == nil {
+		log.WithField("ClusterGUID", kv.Value).Info("Assigned cluster GUID")
+	} else {
+		if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+			log.WithError(err).WithField("ClusterGUID", kv.Value).Warn("Failed to assign cluster GUID")
```

Comment:
wouldn't this throw an error if calico-node simply reboots and calls `EnsureInitialized` to check the connection because ClusterGUID was already `Create`d from the first bootup?

Reply:
eh?  this doesn't return an error if the ClusterGUID was already created.

Reply:
oooo nvm, it's `!ok` case. Ignore that.